### PR TITLE
Add support for custom roundtripper

### PR DESCRIPTION
**Pull request by milinddethe15**
* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes
Added a parameter in `NewBucket()` to pass custom roundtripper

## Verification
Created tests in which a `always error returning` roundtripper is passed to NewBucket() and tested whether error is caught or not.

### CODE REVIEW

File: providers/azure/helpers.go [hunk 1]
```diff
@@ -20,9 +20,17 @@ import (
...
+	// otherwise build the default transport.
+	var rt http.RoundTripper
+	if conf.HTTPConfig.Transport != nil {
+		rt = conf.HTTPConfig.Transport
+	} else {
+		var err error
+		rt, err = exthttp.DefaultTransport(conf.HTTPConfig)
+		if err != nil {
+			return nil, err
+		}
```
Comment (saswatamcode) [inline providers/azure/helpers.go:33]: Let's do only if
```suggestion
	rt, err := exthttp.DefaultTransport(conf.HTTPConfig)
	if err != nil {
		return nil, err
	}
	if conf.HTTPConfig.Transport != nil {
		rt = conf.HTTPConfig.Transport
	}
		
```

File: providers/s3/s3_test.go [hunk 2]
```diff
@@ -458,7 +459,29 @@ func TestParseConfig_CustomStorageClass(t *testing.T) {
...
 	cfg := DefaultConfig
 	cfg.Endpoint = endpoint
-	bkt, err := NewBucketWithConfig(log.NewNopLogger(), cfg, "test")
+	bkt, err := NewBucketWithConfig(log.NewNopLogger(), cfg, "test", nil)
 	testutil.Ok(t, err)
 	testutil.Equals(t, "", bkt.storageClass)
 }
+
+// ErrorRoundTripper is a custom RoundTripper that always returns an error
+type ErrorRoundTripper struct {
```
Comment (saswatamcode) [inline providers/s3/s3_test.go:468]: Let's define this in single place and reuse. Maybe errutil

File: providers/azure/azure_test.go [hunk 3]
```diff
@@ -20,7 +23,7 @@ type TestCase struct {
 }
 
 var validConfig = []byte(`storage_account: "myStorageAccount"
-storage_account_key: "abc123"
+storage_account_key: "bXlTdXBlclNlY3JldEtleTEyMyFAIw=="
```
Comment (saswatamcode) [inline providers/azure/azure_test.go:26]: Does this need to be base64 encoded?

File: providers/bos/bos.go [hunk 4]
```diff
@@ -66,6 +66,7 @@ func parseConfig(conf []byte) (Config, error) {
 
 // NewBucket new bos bucket.
 func NewBucket(logger log.Logger, conf []byte, component string) (*Bucket, error) {
+	//TODO: Add support for custom roundtripper.
```
Comment (saswatamcode) [inline providers/bos/bos.go:69]: Let's mention this PR or original issue link as well?

File: providers/oss/oss_test.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+	"testing"
+
+	"github.com/efficientgo/core/testutil"
+	"github.com/go-kit/log"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/objstore/errutil"
+)
+
+func TestNewBucketWithErrorRoundTripper(t *testing.T) {
+
```
Comment (saswatamcode) [inline providers/oss/oss_test.go:14]:
```suggestion
```

File: providers/oci/oci_test.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+	"github.com/efficientgo/core/testutil"
+	"github.com/go-kit/log"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/objstore/errutil"
+	"gopkg.in/yaml.v2"
+)
+
+func TestNewBucketWithErrorRoundTripper(t *testing.T) {
+
```
Comment (saswatamcode) [inline providers/oci/oci_test.go:14]:
```suggestion
```

File: providers/oci/oci_test.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	mockPrivateKey := `-----BEGIN RSA PRIVATE KEY-----
```
Comment (saswatamcode) [inline providers/oci/oci_test.go:15]: Let's make this a constant

File: providers/oss/oss.go [hunk 8]
```diff
@@ -158,22 +158,27 @@ func (b *Bucket) Attributes(ctx context.Context, name string) (objstore.ObjectAt
...
 // NewBucketWithConfig returns a new Bucket using the provided oss config struct.
-func NewBucketWithConfig(logger log.Logger, config Config, component string) (*Bucket, error) {
+func NewBucketWithConfig(logger log.Logger, config Config, component string, rt http.RoundTripper) (*Bucket, error) {
 	if err := validate(config); err != nil {
 		return nil, err
 	}
-
 	client, err := alioss.New(config.Endpoint, config.AccessKeyID, config.AccessKeySecret)
+	if rt != nil {
+		// custom RoundTripper
```
Comment (saswatamcode) [inline providers/oss/oss.go:176]: Fine to remove the comment here 
```suggestion
```

File: providers/oss/oss_test.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+	var config Config
+	config.Endpoint = "http://test.com/"
+	config.AccessKeyID = "123"
+	config.AccessKeySecret = "123"
+	config.Bucket = "test"
+	config.AccessKeySecret = "123"
```
Comment (saswatamcode) [inline providers/oss/oss_test.go:20]: Let's assign in struct directly?

File: providers/obs/obs.go [hunk 10]
```diff
@@ -75,11 +75,11 @@ type Bucket struct {
 }
 
 func NewBucket(logger log.Logger, conf []byte) (*Bucket, error) {
+	//TODO: Add support for custom roundtripper.
```
Comment (saswatamcode) [inline providers/obs/obs.go:78]: Same here, let's mention link in TODO

File: providers/gcs/gcs_test.go [hunk 11]
```diff
@@ -157,3 +159,25 @@ http_config:
 		})
 	}
 }
+
+func TestNewBucketWithErrorRoundTripper(t *testing.T) {
+	// Create an error RoundTripper.
```
Comment (saswatamcode) [inline providers/gcs/gcs_test.go:164]: No need for the comment here
```suggestion
```

File: providers/azure/helpers.go [hunk 12]
```diff
@@ -20,10 +20,16 @@ import (
 const DirDelim = "/"
 
 func getContainerClient(conf Config) (*container.Client, error) {
-	dt, err := exthttp.DefaultTransport(conf.HTTPConfig)
+	// Check if a roundtripper has been set in the config
+	// otherwise build the default transport.
```
Comment (saswatamcode) [inline providers/azure/helpers.go:24]: I feel like this is discernable from the code itself so we can probably skip comments.
```suggestion
```

**Review by saswatamcode — APPROVED** [review]
Generally LGTM!
Just some minor nits

File: providers/azure/azure_test.go (hunk 3 above)
Comment (milinddethe15) [reply providers/azure/azure_test.go:26]: yes, otherwise it would give error.

**Review by GiedriusS — APPROVED** [review]

File: providers/s3/s3.go [hunk 13]
```diff
@@ -201,7 +201,7 @@ func (s *overrideSignerType) Retrieve() (credentials.Value, error) {
 }
 
 // NewBucketWithConfig returns a new Bucket using the provided s3 config values.
-func NewBucketWithConfig(logger log.Logger, config Config, component string) (*Bucket, error) {
+func NewBucketWithConfig(logger log.Logger, config Config, component string, rt http.RoundTripper) (*Bucket, error) {
```
Comment (yeya24) [inline providers/s3/s3.go:204]: I might be too late to this. But if we already have `config.HTTPConfig.Transport`, why do we still introduce a new `rt` parameter?

**Pull request merged**